File: src/1406_StoneGameIII.cpp
#include <local_leetcode.hpp>

class Solution {
  public:
    string stoneGameIII(vector<int>& stoneValue) {
      const int n = stoneValue.size();

      int prefix[n + 3];

      prefix[0] = 0;
      for (int i = 0; i < n; ++i) 
        prefix[i + 1] = prefix[i] + stoneValue[i];
      prefix[n + 2] = prefix[n + 1] = prefix[n];

      int dp[n + 3];
      dp[n + 2] = dp[n + 1] = dp[n] = 0;

      for (int i = n - 1; i >= 0; --i) {
        int res = INT_MIN;

        for (int j = 1; j <= 3; ++j)
          res = max(res, prefix[i + j] - prefix[i] - dp[i + j]);

        dp[i] = res;
      }

      return dp[0] == 0 ? "Tie" : (dp[0] > 0 ? "Alice" : "Bob");
    }
};


int main() {
  EXECS(Solution::stoneGameIII);
  return 0;
}

File: src/1406_StoneGameIII.java
import java.util.Arrays;

class StoneGameIII {
  public static void main(String[] args) {
    int[] stones = new int[] {-1,-2,-3};
    StoneGameIII obj = new StoneGameIII();
    System.out.println(obj.stoneGameIII(stones));
  }
  int N;
  int[] stones, a, b;
  public String stoneGameIII(int[] stoneValue) {
    stones = stoneValue;
    N = stoneValue.length;
    a = new int[N];
    b = new int[N];
    Arrays.fill(a, Integer.MIN_VALUE);
    Arrays.fill(b, Integer.MAX_VALUE);

    int alice = dfs(0, true);
    int bob = Arrays.stream(stoneValue).sum() - alice;
    return alice==bob ? "Tie" : (alice>bob ? "Alice" : "Bob");
  }

  int dfs(int idx, boolean alice) {
    if (idx>=N) return 0;
    if (alice) {
      if (a[idx]!=Integer.MIN_VALUE)
        return a[idx];
      int res=Integer.MIN_VALUE, acc=0;
      for (int k=0; k<Math.min(N-idx, 3); ++k) {
        acc += stones[idx+k];
        res = Math.max(res, acc + dfs(idx+k+1, !alice));
      }
      a[idx] = res;
      return res;
    }
    else {
      if (b[idx]!=Integer.MAX_VALUE)
        return b[idx];
      int res = Integer.MAX_VALUE;
      for (int k=0; k<Math.min(N-idx, 3); ++k) 
        res = Math.min(res, dfs(idx+k+1, !alice));
      b[idx] = res;
      return res;
    }
  }
}